src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,980 @@
+        DSTOffset::TypeInfo::Type dstList = TimeSynchronizationServer::Instance().GetDSTOffset();
+        for (const auto & dstOffset : dstList)
+        {
+            ReturnErrorOnFailure(encoder.Encode(dstOffset));
+        }
+
+        return CHIP_NO_ERROR;
+    });
+
+    return err;
+}
+
+CHIP_ERROR TimeSynchronizationAttrAccess::ReadLocalTime(EndpointId endpoint, AttributeValueEncoder & aEncoder)
+{
+    DataModel::Nullable<uint64_t> localTime;
+    CHIP_ERROR err = TimeSynchronizationServer::Instance().GetLocalTime(endpoint, localTime);
+    err            = aEncoder.Encode(localTime);
+    return err;
+}
+
+CHIP_ERROR TimeSynchronizationAttrAccess::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    if (aPath.mClusterId != Id)
+    {
+        return CHIP_ERROR_INVALID_PATH_LIST;
+    }
+
+    switch (aPath.mAttributeId)
+    {
+    case UTCTime::Id: {
+        System::Clock::Microseconds64 utcTimeUnix;
+        uint64_t chipEpochTime;
+
+        VerifyOrReturnError(System::SystemClock().GetClock_RealTime(utcTimeUnix) == CHIP_NO_ERROR, aEncoder.EncodeNull());
+        VerifyOrReturnError(UnixEpochToChipEpochMicro(utcTimeUnix.count(), chipEpochTime), aEncoder.EncodeNull());
+        return aEncoder.Encode(chipEpochTime);
+    }
+    case Granularity::Id: {
+        return aEncoder.Encode(TimeSynchronizationServer::Instance().GetGranularity());
+    }
+    case TrustedTimeSource::Id: {
+        return ReadTrustedTimeSource(aPath.mEndpointId, aEncoder);
+    }
+    case DefaultNTP::Id: {
+        return ReadDefaultNtp(aPath.mEndpointId, aEncoder);
+    }
+    case TimeZone::Id: {
+        return ReadTimeZone(aPath.mEndpointId, aEncoder);
+    }
+    case DSTOffset::Id: {
+        return ReadDSTOffset(aPath.mEndpointId, aEncoder);
+    }
+    case TimeZoneListMaxSize::Id: {
+        uint8_t max = CHIP_CONFIG_TIME_ZONE_LIST_MAX_SIZE;
+        return aEncoder.Encode(max);
+    }
+    case DSTOffsetListMaxSize::Id: {
+        uint8_t max = CHIP_CONFIG_DST_OFFSET_LIST_MAX_SIZE;
+        return aEncoder.Encode(max);
+    }
+    case LocalTime::Id: {
+        return ReadLocalTime(aPath.mEndpointId, aEncoder);
+    }
+    default: {
+        break;
+    }
+    }
+
+    return err;
+}
+} // anonymous namespace
+
+bool emberAfTimeSynchronizationClusterSetUTCTimeCallback(
+    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
+    const chip::app::Clusters::TimeSynchronization::Commands::SetUTCTime::DecodableType & commandData)
+{
+    const auto & utcTime     = commandData.UTCTime;
+    const auto & granularity = commandData.granularity;
+    const auto & timeSource  = commandData.timeSource;
+
+    auto currentGranularity = TimeSynchronizationServer::Instance().GetGranularity();
+    if (granularity < GranularityEnum::kNoTimeGranularity || granularity > GranularityEnum::kMicrosecondsGranularity)
```

Comment:
This would be more forward-compatible as:
```suggestion
    if (granularity == GranularityEnum::kUnknownEnumValue)
```
Followup is fine.

Reply:
Tracked in https://github.com/project-chip/connectedhomeip/issues/27575, resolving this comment for now, further followup on the issue.